Adding TTL to CWeapon

seedship opened the pull request:
Added option to set `ttl` in `SetUnitWeaponState`, and added a `ttl` field into `CWeapon`. This overrides the original `weaponDef->flighttime` so games can set weapon ttl per weapon.

Untested, idk how to test -- @GoogleFrog any tips?

seedship commented:
Addresses https://github.com/beyond-all-reason/spring/issues/1683

seedship commented on rts/Sim/Weapons/Weapon.h:
```diff
@@ -151,6 +151,7 @@ class CWeapon : public CObject
 	int nextSalvo;                          // when the next shot in the current salvo will fire
 	int salvoLeft;                          // number of shots left in current salvo
 	int salvoWindup;                        // delay before first shot (in frames)
+	int ttl;                                // flight time for missile type weapons
```
Renamed the comment to say projectile type weapons. It does set the `ttl` field in the projectile params.

seedship commented:
It seems Starburst weapons/projectiles are complicated and do their own thing. I will leave them alone in this PR.

saurtron commented on rts/Sim/Weapons/StarburstLauncher.cpp:
```diff
@@ -39,9 +39,10 @@ void CStarburstLauncher::FireImpl(const bool scriptCall)
 	params.end = currentTargetPos;
 	params.speed = speed;
 	params.error = aimError;
-	params.ttl = 200; //???
```
I think you can make this pass ttl on and then make StarburstProjectile.cpp not overwrite it from weaponDef. That way ttl should also be controllable for starburst.

saurtron commented on rts/Sim/Weapons/Cannon.cpp:
```diff
@@ -110,27 +110,27 @@ void CCannon::FireImpl(const bool scriptCall)
 	launchDir.SafeNormalize();
 
-	int ttl = 0;
+	int thisTtl = 0;
```
My eyes bleed xD, but not the only place that writes it (Ttl) like that (vs TTL) xD. Also maybe consider myTTL or myTtl instead of thisTtl/thisTTL, I think it's more in line with style around these files.

Anyways not really an issue since it's more of a personal preference thing, you can choose any you want from thisTtl/thisTTL/myTtl/myTTL

seedship replied:
Hmmm yeah I think in camelcase abbreviations should be treated as a single word, rather than in all caps

sprunk commented on rts/Lua/LuaSyncedCtrl.cpp:
```diff
@@ -2206,6 +2206,10 @@ static bool SetSingleUnitWeaponState(lua_State* L, CWeapon* weapon, int index)
 		} break;
 
+		case hashString("ttl"): {
+			weapon->ttl = lua_toint(L, index + 1);
```
```suggestion
			weapon->ttl =  (int) (lua_tofloat(L, index + 1) * GAME_SPEED);
```

saurtron replied:
This can make it a bit awkward if a specific number of frames is desired, otherwise the SetSingleUnitWeaponState key should maybe be called flightTime since ttl is in frames for projectiles SetProjectileTimeToLive (and in that case I think having perfect control of the ttl frames is actually important).

sprunk replied:
You can control frames perfectly by multiplying by frames per second. Calling it `flightTime` sounds fine.

saurtron replied:
Weapondef also is in frames, imo better to have all in frames for consistency.

sprunk replied:
No it's not.

saurtron replied:
Right, perfect then, lets call it `flightTime` and make it in seconds.

seedship commented:
Going on vacation for a week, will address the comments when I get back.

seedship commented:
Addressed comments:

- Made starburst take in ttl from projectile params
- Made TTL in seconds

Verified the changes by using the gadget in the description and verified missiles visually run out of fuel after 1 second

sprunk reviewed (approved):
Haven't tested yet but looks good.

badosu reviewed (changes requested):
- Missing the corresponding entry for [`LuaSyncedRead::GetUnitWeaponState`](https://github.com/beyond-all-reason/spring/blob/198c84e8b3c2b57490d790dddcb7efd69b11c025/rts/Lua/LuaSyncedRead.cpp#L5132).
- Please add the `ttl` `@field` to the [`WeaponState`](https://github.com/beyond-all-reason/spring/blob/198c84e8b3c2b57490d790dddcb7efd69b11c025/rts/Lua/LuaSyncedCtrl.cpp#L2205) type. See [the doc page](https://beyond-all-reason.github.io/spring/development/documenting-lua#annotations) for questions on documenting lua types.

sprunk commented on rts/Lua/LuaSyncedCtrl.cpp:
```diff
@@ -2243,6 +2243,10 @@ static bool SetSingleUnitWeaponState(lua_State* L, CWeapon* weapon, int index)
 		} break;
 
+		case hashString("ttl"): {
+			weapon->ttl = (int) (lua_toint(L, index + 1) * GAME_SPEED);
```
Should be float since now you can't pass e.g. 0.9 since it gets rounded to 0.
```suggestion
			weapon->ttl = (int) (lua_tonumber(L, index + 1) * GAME_SPEED);
```

sprunk commented on rts/Lua/LuaSyncedCtrl.cpp:
```diff
@@ -2243,6 +2243,10 @@ static bool SetSingleUnitWeaponState(lua_State* L, CWeapon* weapon, int index)
 		} break;
 
+		case hashString("ttl"): {
+			weapon->ttl = (int) (lua_toint(L, index + 1) * GAME_SPEED);
```
Should be float since now you can't pass e.g. 0.9 since it gets rounded to 0.
```suggestion
			weapon->ttl = (int) (lua_tonumber(L, index + 1) * GAME_SPEED);
```

seedship replied:
Whoops, my bad, thanks!

sprunk commented on rts/Lua/LuaSyncedCtrl.cpp:
```diff
@@ -2313,6 +2314,10 @@ static bool SetSingleUnitWeaponState(lua_State* L, CWeapon* weapon, int index)
 		} break;
 
+		case hashString("ttl"): {
+			weapon->ttl = (int) (lua_tonumber(L, index + 1) * GAME_SPEED);
```
All other tags are also just truncated.

badosu commented on rts/Lua/LuaSyncedCtrl.cpp:
```diff
@@ -2240,6 +2240,7 @@ int LuaSyncedCtrl::SetUnitStockpile(lua_State* L)
  * @field forceAim integer?
  * @field avoidFlags integer?
  * @field collisionFlags integer?
+ * @field ttl integer?
```
Please add a description, e.g. ` * @field ttl integer? The time to live in bla`

badosu reviewed (approved):
My concerns were mostly resolved. I posted a comment asking for the field description on the docs.

I'm pre-approving this PR as the general implementation looks solid to me, apart from adjustments required from other reviewers.

seedship commented on rts/Lua/LuaSyncedRead.cpp:
```diff
@@ -80,6 +80,8 @@
 #include <cctype>
 #include <type_traits>
 
+#include <iostream>
```
Whoops sorry was left over from debugging
